distributed/scheduler.py

```diff
@@ -654,6 +650,96 @@ def _to_dict_no_nest(self, *, exclude: Container[str] = ()) -> dict[str, Any]:
             members=True,
         )
 
+    @property
+    def scheduler(self):
+        assert self.scheduler_ref
+        s = self.scheduler_ref()
```

Comment:
To be honest, the more I look at this, I'm considering moving these functions (or at least the parts that require access to a `SchedulerState` back up the hierarchy into the scheduler state. Since we need access to the `SchedulerState` as a whole, it seems that this would be the better correct root object to handle those operations. This would also avoid problems like acquiring resources when using `Scheduler._add_to_processing` but not doing so in `WorkerState.add_to_processing`. It's not clear which of these methods should be used by others such as `WorkStealing` and I'm increasingly convinced that it should not be the `WorkerState`-based ones.

Reply:
This is a bit of an elaborate change, so I think it might make sense to extract it into an individual PR to avoid blocking this one and litter some technically unrelated refactoring changes into this.